core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -211,7 +211,13 @@ public TransportClientNodesService removeTransportAddress(TransportAddress trans
     }
 
     public <Response> void execute(NodeListenerCallback<Response> callback, ActionListener<Response> listener) {
-        List<DiscoveryNode> nodes = this.nodes;
+        List<DiscoveryNode> nodes;
+        synchronized (mutex) {
```

Comment:
FYI - Node sampling is done under the mutex - see [here](https://github.com/jasontedor/elasticsearch/blob/e34961f2881a8cdf461bf4ab34c1e66b2968d97b/core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java#L315) . This is blocking and can thus delay sending requests which is a no no.  

IMO it will be better to go with the suggestion of @ywelsch rather than go and adapt sniffing to be async.